### DISCUSSION TITLE
Docker cli login potential security issue or not?

### DISCUSSION THREAD

**Discussion by aleks-ivanov** [community/community#26826] [Actions]
<p>Login to DockerHub in a GiHA pipeline using the Docker CLI:</p>
<pre><code class="lang-auto">Run DOCKER_USER=***
  DOCKER_USER=***
  DOCKER_PASSWORD=***
  
  echo $DOCKER_PASSWORD | docker login --username $DOCKER_USER \
  --password-stdin
  shell: /usr/bin/bash -e {0}
WARNING! Your password will be stored unencrypted in /home/runner/.docker/config.json.
Login Succeeded
</code></pre>
<p>The *** represent ${{ secrets.DOCKER_USER }} and ${{ secrets.DOCKER_PASSWORD }} respectively.</p>
<p>So, I followed the warning, found the file and printed it in the console:</p>
<pre><code class="lang-auto">{
        "auths": {
                "https://index.docker.io/v1/": {
                        "auth": "long-string-of-random-characters"
                }
        }
}
</code></pre>
<p>So here I found that the password to my Dhub account was indeed encrypted by GitHub’s Secrets module and appended to this file encrypted, unlike the warning in the log indicated.</p>
<p>Now what I am wondering, if the password is encrypted by the Secrets module is there really a security issue with that value being appended to the <code>config.json</code> file, since the only way someone would have access to that file in that state is to that someone having access to my account in the first place ?</p>
<p>My conclusion is <code>No.</code>, but I would like to have a second/third/Nth opinion of someone more experienced than moi.</p>

**Comment by airtower-luna** [comment, marked as answer]
<aside class="quote no-group" data-username="aleks-ivanov" data-post="1" data-topic="202611">
<div class="title">
<div class="quote-controls"></div>
 aleks-ivanov:</div>
<blockquote>
<p>So here I found that the password to my Dhub account was indeed encrypted by GitHub’s Secrets module and appended to this file encrypted, unlike the warning in the log indicated.</p>
</blockquote>
</aside>
<p>That conclusion is unfortunately wrong, try base64-decoding that <code>long-string-of-random-characters</code> thing. :wink:</p>
<p>If someone could get access to the runner while your credentials are in there they could use them. It’s up to you whether you consider that an unacceptable risk. It is common practice to ensure that the credentials are deleted after use, even though the GitHub-hosted runner VMs are discarded after the job.</p>

**Comment by aleks-ivanov** [comment]
<p>That’s what I was unsure about, thanks for clarifying.</p>
<p>What are ways that someone could access the runner during execution, other than having access to my GH account ?</p>

**Comment by airtower-luna** [comment]
<p>If you use GitHub-hosted runners, no-one should be able to access them, assuming that:</p>
<ul>
<li>GitHub’s security systems work as designed.</li>
<li>You don’t do anything in the workflow to allow outside access or exfiltrate secrets, or that has security issues that allow others to do so.</li>
</ul>
<p>If you use self-hosted runners, it’s up to you to secure them.</p>

**Comment by jhsu802701** [comment]
Is there a more secure solution for allowing GitHub Workflows to push Docker images?  This is basically the same solution I'm using now, and it does work.  However, I'd like to switch to a better solution, and I cannot replace something with nothing.

**Reply by airtower-luna** [reply]
It depends on the registry you're pushing to. If it's `ghcr.io`, you can use the `GITHUB_TOKEN`. If the registry supports it, you could set up [OpenID for authentication](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-cloud-providers). In both cases you'd have short-lived tokens, which would limit damage in case someone manages to steal one.

**Reply by jhsu802701** [reply]
Yes, I'm using ghcr.io.  My build.yml file is at https://github.com/rubyonracetracks/docker-debian-bullseye-min-stage1/blob/main/.github/workflows/build.yml .  Is there any example of a repository that uses the more secure solution that you've suggested here?  My way cuts it for getting the job done but not for doing it in the most secure manner.

**Reply by airtower-luna** [reply]
You're already using the short-lived token, so there are mostly minor things you could do better:

* [Limit workflow permissions](https://github.com/airtower-luna/hello-ghcr/blob/27eb07f88f5d134e3b55e6e8f332f060fb82d0c8/.github/workflows/build.yaml#L12-L15) to what's actually needed (this is the one major thing).
* Set the environment variable with the token [only for the login step](https://github.com/airtower-luna/hello-ghcr/blob/27eb07f88f5d134e3b55e6e8f332f060fb82d0c8/.github/workflows/build.yaml#L62-L69), not the whole job. That way steps that don't need it won't be able to access it.
* Add a [logout step](https://github.com/airtower-luna/hello-ghcr/blob/27eb07f88f5d134e3b55e6e8f332f060fb82d0c8/.github/workflows/build.yaml#L76-L79) with `if: always()`, to make sure the token is removed from disk before the job ends.

I've linked examples from one of my own repositories. I don't see anything in your workflow that'd need the `security-events` permission, so you can remove that, too. :smile_cat:  

As for documentation, [Automatic token authentication](https://docs.github.com/en/actions/security-guides/automatic-token-authentication) and [Security hardening](https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions) are probably the most interesting parts for your case.

**Reply by jhsu802701** [reply]
Thanks for the example.  I'll make your suggested improvements, but I see from the GitHub actions that you're still getting the following warning message:
```
WARNING! Your password will be stored unencrypted in /home/runner/.docker/config.json.
Configure a credential helper to remove this warning. See
https://docs.docker.com/engine/reference/commandline/login/#credentials-store
```

Have you figured how to use the suggested credential helper?  I wish I could see a GitHub Workflows example of this in action.

**Reply by airtower-luna** [reply]
I have, but it doesn't make sense to do inside the Actions runner in my opinion. The warning is mostly meant for persistent systems, like: "Hey, the credentials will end up on disk in plain text!" On my work laptop I have set up Docker to use `pass`, which means the credentials get stored with PGP encryption. Technically you could do that in the Actions runner too, but you wouldn't gain anything, because then you'd need to worry about how to protect the key (or its passphrase) instead. :sweat_smile: 

Limiting the credential access to the steps that actually need them should be reasonably safe.
